### PR TITLE
[DataFrame] Sample implement

### PR DESCRIPTION
**Pull request by osalpekar**
Implements `sample()` function for ray dataframes

### CODE REVIEW

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -2801,9 +2801,150 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+        else:
+            # randomly select labels from chosen axis
+            samples = np.random.choice(a=axis_labels, size=n,
+                                       replace=replace, p=weights)
+
+        part_ind_tuples = [partition_metadata[sample]
+                           for sample in samples]
+
+        if axis == 1:
+            new_cols = [_deploy_func.remote(lambda df: df.iloc[:, [tup[1]]],
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Can you add a comment here about the logic. What is `tup[0]` and `tup[1]`, etc.

**Review by devin-petersohn — COMMENTED** [review]
The only thing I would add is some extra comments when the logic gets a bit hard to follow. Looks great!

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -2801,9 +2801,158 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+            else 0
+
+        if axis == 0:
+            axis_length = len(self._row_metadata)
+        else:
+            axis_length = len(self._col_metadata)
+
+        if weights is not None:
+
+            # If a series, align with frame
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: I would prefer for the comments to say why something was done, rather than what. I think that if you say why it's important to align with frame, that would be more helpful in reading the code..

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -2801,9 +2801,158 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+            if isinstance(weights, pd.Series):
+                weights = weights.reindex(self.axes[axis])
+
+            # Strings acceptable if a dataframe and axis = 0
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Same comment as above.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -2801,9 +2801,158 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+                                 "same length")
+
+            if (weights == np.inf).any() or (weights == -np.inf).any():
+                raise ValueError("weight vector may not include `inf` values")
+
+            if (weights < 0).any():
+                raise ValueError("weight vector many not include negative "
+                                 "values")
+
+            # If has nan, set to zero.
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Same comment as above.

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -2801,9 +2801,158 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+            weights = weights.fillna(0)
+
+            # Renormalize if don't sum to 1
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Same comment as above.

File: python/ray/dataframe/dataframe.py [hunk 6]
```diff
@@ -2801,9 +2801,158 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+            if weights.sum() != 1:
+                if weights.sum() != 0:
+                    weights = weights / weights.sum()
+                else:
+                    raise ValueError("Invalid weights: weights sum to zero")
+
+            weights = weights.values
+
+        if n is None and frac is None:
+            # default to n = 1 if n and frac are None
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Same comment as above.

**Review by devin-petersohn — COMMENTED** [review]
I didn't put it everywhere, but in a few of the comments, there is only a description of what is happening, but that should be evident from the code itself. I would prefer the comments to say why something is the way that it is.

File: python/ray/dataframe/dataframe.py [hunk 7]
```diff
@@ -2801,9 +2801,158 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+
+            if (weights < 0).any():
+                raise ValueError("weight vector many not include negative "
+                                 "values")
+
+            # If has nan, set to zero.
+            weights = weights.fillna(0)
+
+            # Renormalize if don't sum to 1
+            if weights.sum() != 1:
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Sum weights once and use the variable for these two if statements and rebalance

**Review by p-yang — CHANGES_REQUESTED** [review]

File: python/ray/dataframe/dataframe.py [hunk 8]
```diff
@@ -2801,9 +2801,166 @@ def rtruediv(self, other, axis='columns', level=None, fill_value=None):
...
+            # weights cannot be NaN when sampling, so we must set all nan
+            # values to 0
+            weights = weights.fillna(0)
+
+            # If passed in weights are not equal to 1, renormalize them
+            # otherwise numpy sampling function will error
+            weights_sum = weights.sum()
+            if weights_sum != 1:
+                if weights_sum != 0:
+                    weights = weights / weights.sum()
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: `weights.sum()` -> `weights_sum`

**Review by devin-petersohn — APPROVED** [review]
